### PR TITLE
BUG 1918920: requeue secret update event when the cluster is being ce…

### PR DESCRIPTION
**Pull request by huikang**
…rtRedeploy

### Description
In current implementation, when two secret update events occur in a short amount of time, the second event will not cause the cluster to be redeployed. This is because the cluster node state has already set to `nodeStatus.UpgradeStatus.ScheduledForCertRedeploy = corev1.ConditionTrue` by the first event. This PR adds the check of current cluster status in the secret reconciliation: requeue the event when the cluster is being `ScheduledForCertRedeploy`

/cc @ewolinetz 
/assign @periklis 

/cherry-pick release-5.0

### Links
- Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1918920

### CODE REVIEW

**Review by periklis — COMMENTED** [review]
/lgtm

File: internal/k8shandler/reconciler.go [hunk 1]
```diff
@@ -77,12 +84,12 @@ func SecretReconcile(requestCluster *elasticsearchv1.Elasticsearch, requestClien
...
 
 	if retryErr != nil {
-		return kverrors.Wrap(retryErr, "failed to update status for cert redeploys",
+		return false, kverrors.Wrap(retryErr, "failed to update status for cert redeploys",
 			"cluster", requestCluster.Name,
 			"retries", nretries)
 	}
 
-	return nil
+	return false, nil
```
Comment (ewolinetz) [inline internal/k8shandler/reconciler.go:92]: can we actually flip this? it seems odd that we'd return `false` for a successful operation

**Comment by huikang** [comment]
@ewolinetz, the problem still exists after applying the PR. The reason is that the nodes' status is always  `scheduledCertRedeploy: 'True'` (see the following node status) after `"unable to complete full cluster restart"`.

```
nodes:
    - conditions:
        - lastTransitionTime: '2021-03-07T17:19:19Z'
          reason: Error
          status: 'True'
          type: ElasticsearchContainerTerminated
        - lastTransitionTime: '2021-03-07T17:19:19Z'
          reason: Error
          status: 'True'
          type: ProxyContainerTerminated
      deploymentName: elasticsearch-cdm-qux9b8lm-1
      upgradeStatus:
        scheduledCertRedeploy: 'True'
```

```
nodes:
- deploymentName: elasticsearch-cdm-qux9b8lm-1
    upgradeStatus:
    scheduledCertRedeploy: 'True'
```

When I tried resetting the ``scheduledCertRedeploy` to `False`; it will be overwritten to `True` by the managing CLO instance. Could you provide any guidance? Thanks.

**Comment by ewolinetz** [comment]
@huikang let me think on this a little... i have an idea but i'm not convinced yet that it wouldn't have the same issue that we currently face.

**Comment by huikang** [comment]
@ewolinetz , sure. I found out that the root cause is that at the end of the first cert redeploy es client can not connect to the cluster due to the second secret change in waitAllNodesRejoinAndSetAllShards(), so post() will fail and postSignaler can not be called to reset the node status.

File: internal/k8shandler/clusterrestart.go [hunk 2]
```diff
@@ -409,8 +409,9 @@ func (r *Restarter) setClusterConditions(updateStatus func()) {
 
 	r.mainCondition = func() bool {
 		return containsClusterCondition(api.Restarting, v1.ConditionTrue, r.clusterStatus) &&
-			containsClusterCondition(api.UpdatingESSettings, v1.ConditionFalse, r.clusterStatus) &&
-			containsClusterCondition(api.Recovering, v1.ConditionFalse, r.clusterStatus)
+			containsClusterCondition(api.UpdatingESSettings, v1.ConditionFalse, r.clusterStatus)
+		// &&
+		// containsClusterCondition(api.Recovering, v1.ConditionFalse, r.clusterStatus)
```
Comment (ewolinetz) [inline internal/k8shandler/clusterrestart.go:414]: i think we should revert the changes in this file... i'm thinking that we actually want to blank things out in the secret controller in the case where we are already mid restart (so we can force it to restart the cert restart process)

**Comment by huikang** [comment]
@ewolinetz , I have updated the PR with the following changes 1) reset cert redeploy in the secret controller, and 2) flip the return value of the reconciler. I tested the PR successfully in my local cluster.

**Comment by ewolinetz** [comment]
@huikang please squash, otherwise :+1:
